go/vault/api/vault.go

```diff
@@ -0,0 +1,135 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/oasisprotocol/oasis-core/go/common/prettyprint"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+)
+
+var _ prettyprint.PrettyPrinter = (*Authority)(nil)
+
+// State is the vault state.
+type State uint8
+
+const (
+	StateActive    = 0
+	StateSuspended = 1
+)
+
+// Vault contains metadata about a vault.
+type Vault struct {
+	// CreatedBy is the address of the vault creator.
+	CreatedBy staking.Address `json:"created_by"`
+	// CreatedNonce is the nonce of the vault creator's account at time of creation.
+	CreatedNonce uint64 `json:"created_nonce"`
+	// State is the vault state.
+	State State `json:"state"`
+	// Nonce is the nonce to use for the next action.
+	Nonce uint64 `json:"nonce,omitempty"`
+
+	// AdminAuthority specifies the vault's admin authority.
+	AdminAuthority Authority `json:"admin_authority"`
+	// SuspendAuthority specifies the vault's suspend authority.
+	SuspendAuthority Authority `json:"suspend_authority"`
+}
+
+// NewVaultAddress returns the address for the vault.
+func NewVaultAddress(createdBy staking.Address, createdNonce uint64) staking.Address {
+	return staking.NewModuleAddress(ModuleName, fmt.Sprintf("vault.%s.%d", createdBy, createdNonce))
+}
+
+// Address returns the address for the vault.
+func (v *Vault) Address() staking.Address {
+	return NewVaultAddress(v.CreatedBy, v.CreatedNonce)
+}
+
+// IsActive retruns true iff the vault is currently active (processing withdrawals).
+func (v *Vault) IsActive() bool {
+	return v.State == StateActive
+}
+
+// Authorities returns the list of all vault authorities.
+func (v *Vault) Authorities() []*Authority {
```

Comment:
Seems a bit weird putting authorities with different scopes into the same list, without any info about the scope. This should likely return a map or an object which would also contain info with the scope of each authority (e.g. admin, suspend). Or maybe even remove this method if it's rarely needed.

Reply:
This is only used in a single place which is early rejection for action cancellations (when the action itself has not yet been fetched from the database so the exact authority is not yet available). We initially reject any callers that are not part of any authority and in the second phase (when action is available) we re-check and only accept the action-specific authority.

Reply:
Yeah maybe consider just explicitly getting both admin and suspend authorities in the list there. Or making this a private method and only used there. Or keep as it is.

Not a big deal, just seems a bit dangerous having a public `Authorities` method which then returns a mix of authorities with different scopes (and without the scope info).

Reply:
The idea is that if a new authority is added, it would be updated in the same place (e.g. here) and the rest of the code would continue to function as before.